packages/apollo-link-retry/src/retryLink.ts

```diff
@@ -53,52 +55,178 @@ export namespace RetryLink {
+ */
+class RetryableOperation<TValue = any> {
+  private retryCount: number = 0;
+  private values: any[] = [];
+  private error: any;
+  private complete = false;
+  private canceled = false;
+  private observers: ZenObservable.Observer<TValue>[] = [];
+  private currentSubscription: ZenObservable.Subscription = null;
+  private timerId: number;
+
+  constructor(
+    private operation: Operation,
+    private nextLink: NextLink,
+    private retryAfter: RetryLink.IntervalPredicate,
+  ) {}
+
+  /**
+   * Register a new observer for this operation.
+   *
+   * If the operation has previously emitted other events, they will be
+   * immediately triggered for the observer.
+   */
+  subscribe(observer: ZenObservable.Observer<TValue>) {
+    if (this.canceled) {
+      throw new Error(
+        `Subscribing to a retryable link that was canceled is not supported`,
+      );
+    }
+    this.observers.push(observer);
+
+    // If we've already begun, catch this observer up.
+    for (const value of this.values) {
+      observer.next(value);
+    }
+
+    if (this.complete) {
+      observer.complete();
+    } else if (this.error) {
+      observer.error(this.error);
+    }
+  }
+
+  /**
+   * Remove a previously registered observer from this operation.
+   *
+   * If no observers remain, the operation will stop retrying, and unsubscribe
+   * from its downstream link.
+   */
+  unsubscribe(observer: ZenObservable.Observer<TValue>) {
+    const index = this.observers.indexOf(observer);
+    if (index < 0) {
+      throw new Error(
+        `RetryLink BUG! Attempting to unsubscribe unknown observer!`,
+      );
+    }
+    // Note that we are careful not to change the order of length of the array,
+    // as we are often mid-iteration when calling this method.
+    this.observers[index] = null;
+
+    // If this is the last observer, we're done.
+    if (this.observers.every(o => o === null)) {
+      this.cancel();
+    }
+  }
+
+  /**
+   * Start the initial request.
+   */
+  start() {
+    if (this.currentSubscription) return; // Already started.
+
+    this.try();
+  }
+
+  /**
+   * Stop retrying for the operation, and cancel any in-progress requests.
+   */
+  cancel() {
+    if (this.currentSubscription) {
+      this.currentSubscription.unsubscribe();
+    }
+    clearTimeout(this.timerId);
+    this.timerId = null;
+    this.currentSubscription = null;
+    this.canceled = true;
+  }
+
+  private try() {
+    this.currentSubscription = this.nextLink(this.operation).subscribe({
+      next: this.onNext,
+      error: this.onError,
+      complete: this.onComplete,
+    });
+  }
+
+  private onNext = (value: any) => {
+    this.values.push(value);
+    for (const observer of this.observers) {
+      observer.next(value);
+    }
+  };
+
+  private onComplete = () => {
+    for (const observer of this.observers) {
+      try {
+        observer.complete();
+      } catch (error) {}
```

Comment:
whats the reason for the try here?

Reply:
Doh! debugging code that snuck in - removing 👍